rfcs/0031-bazel.md

```diff
@@ -0,0 +1,358 @@
+required, must be complimented by a traditional `configure` file or
+something similar.
+
+## Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+Descriptions of the general benefits of Bazel are easily found with a
+simple websearch, so I'll keep it brief:
+* Reliable, replicable builds.
+* Support for multiple languages.
+* Developed and heavily used by Google; also used by many other major projects.
+* Eliminates the need to separately install/config/build third-party
+  dependencies (apt-get, homebrew, etc.) Bazel will download and build
+  everything.
+* Support for remote caching and execution is (probably) unique.
+
+Alternatives
+* Stick with current situation.  That might work for o(1) internally
+  but is not so attractive for outside volunteers who would have to
+  struggle with the build system.  My guess is that improving the
+  current system would probably be more expensive in the long run than
+  switching to Bazel.
+* Nix may be an alternative, but according to the folks who provide
+  Bazel support for Haskell, Bazel and Nix work well together: [NIX +
+  BAZEL = FULLY REPRODUCIBLE, INCREMENTAL
+  BUILDS](https://www.tweag.io/posts/2018-03-15-bazel-nix.html).
+
+## Prior art
+[prior-art]: #prior-art
+
+Blogposts about switching to Bazel can be found with a little web searching.
+
+I spent considerable effort a few years ago converting a modestly
+sized but somewhat complex C/C++ project
+([OpenOCF](https://github.com/OpenOCF/iochibity)) from SCons to Bazel.
+Though the project was never completed, the switch to Bazel was, and
+was (in my opinion) quite successful.
+
+## Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+* What parts of the design do you expect to resolve through the RFC process before this gets merged?
+
+  * Which platforms and toolchains are to be officially supported.
+
+* What parts of the design do you expect to resolve through the implementation of this feature before merge?
+  * Optimal OCaml/Bazel integration strategy
+
+* What related issues do you consider out of scope for this RFC that could be addressed in the future independently of the solution that comes out of this RFC?
+  * Packaging and distribution.  Bazel just builds stuff; it does not
+    install or package.  But presumably one could write a Bazel
+    package that would do this.
+  * Distributed builds ([Remote Execution](https://docs.bazel.build/versions/master/remote-execution.html))
+  * [Remote caching](https://docs.bazel.build/versions/master/remote-caching.html)
+  * Implementation of feature-testing in Starlark.
+
+#### Known Unknowns
+
+- Will Bazel and the OCaml ecosystem play nice?  Bazel is design as a
+  polyglot system, and already supports a variety of languages, so we
+  have reason to be optimistic about this.
+- Will it really be cheaper/more efficient etc. to rely on one generic
+  build tool, instead of mashing up multiple specialized tools?  Bazel
```

Comment:
You'll most probably discover certain limitations by the underlying tools (like determinism, timestamps, undeclared outputs, dynamic dependencies) while operating in Bazel's stricter environment, and have to wrap / fork / patch those tools to do what you want.